pennylane/data/data_manager/__init__.py

```diff
@@ -0,0 +1,322 @@
+    """
+
+    remote_dataset = Dataset(open_hdf5_s3(s3_url))
+    remote_dataset.write(dest, "a", attributes, overwrite=overwrite)
+
+    del remote_dataset
+
+
+def _download_dataset(
+    data_path: DataPath,
+    dest: Path,
+    attributes: Optional[typing.Iterable[str]],
+    force: bool = False,
+) -> None:
+    """Downloads the dataset at ``data_path`` to ``dest``, optionally downloading
+    only requested attributes."""
+    s3_path = f"{S3_URL}/{data_path}"
+
+    if attributes is not None:
+        _download_partial(s3_path, dest, attributes, overwrite=force)
+        return
+
+    with open(dest, "wb") as f:
+        resp = get(s3_path, timeout=5.0)
+        resp.raise_for_status()
+
+        f.write(resp.content)
+
+
+def load(  # pylint: disable=too-many-arguments
+    data_name: str,
+    attributes: Optional[typing.Iterable[str]] = None,
+    folder_path: Path = Path("./datasets/"),
+    force: bool = False,
+    num_threads: int = 50,
+    cache_dir: Optional[Path] = Path(".cache"),
+    **params: Union[ParamArg, str, List[str]],
+):
+    r"""Downloads the data if it is not already present in the directory and return it to user as a
+    :class:`~pennylane.data.Dataset` object. For the full list of available datasets, please see the
+    `datasets website <https://pennylane.ai/qml/datasets.html>`_.
+
+    Args:
+        data_name (str)   : A string representing the type of data required such as `qchem`, `qpsin`, etc.
+        attributes (list[str]) : An optional list to specify individual data element that are required
+        force (Bool)      : Bool representing whether data has to be downloaded even if it is still present
+        download_dir (str) : Path to the directory used for saving datasets
+        num_threads (int) : The maximum number of threads to spawn while downloading files (1 thread per file)
+        params (kwargs)   : Keyword arguments exactly matching the parameters required for the data type.
+            Note that these are not optional
+
+    Returns:
+        list[:class:`~pennylane.data.Dataset`]
+    """
+    params = format_params(**params)
+
+    folder_path = Path(folder_path)
+    if cache_dir and not Path(cache_dir).is_absolute():
+        cache_dir = folder_path / cache_dir
+
+    foldermap = _get_foldermap()
+
+    data_paths = [data_path for _, data_path in foldermap.find(data_name, **params)]
+
+    dest_paths = [folder_path / data_path for data_path in data_paths]
+
+    for path_parents in set(path.parent for path in dest_paths):
+        path_parents.mkdir(parents=True, exist_ok=True)
+
+    with ThreadPoolExecutor(min(num_threads, len(dest_paths))) as pool:
+        futures = [
+            pool.submit(_download_dataset, data_path, dest_path, attributes, force=force)
+            for data_path, dest_path in zip(data_paths, dest_paths)
+        ]
+        results = wait(futures, return_when=FIRST_EXCEPTION)
+        for result in results.done:
+            if result.exception():
```

Comment:
What is the type/instance of the `result` object here?

Reply:
this is taken from the old Dataset code, and it's [concurrent.futures.wait](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait). It returns a named tuple with 2 elements, `done` and `not_done`. `results.done` is a list of [Futures](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future), and their `exception()` method returns `None` if it completed without error. I never wrote handling for the `not_done` component (because whenever I tried, it was empty), but we've made it this far : )

Reply:
Ahh yes. I just couldn't figure out if it was a list of `Futures`. Using `if result.exception() is not None` might be helpful for clarity here. 
Also, `not_done` is meant to be empty (in the end) because it is supposed to contain the requests that are being handled or are in the queue.

Reply:
that makes sense, but then why does the `not_done` component even exist/get returned by `wait`? 🤔 I'm fine with adding `is not None` 👍